### PR TITLE
v1.3B mark attendance update

### LINKED ISSUE #165: Modify the attendance field to indicate attendance status
For a more comprehensive view of the attendance, we must include attendance statuses in the attendance list chain.

### PR DESCRIPTION
**Pull request by proto-aiken-13**
## Description

Updated the mark attendance command.

## Context

Updated the mark attendance command as per the team's feedback to include a status field and to have better coverage.

Closes #165

## Checklist

- [X] I have self-reviewed my changes.
- [X] I have added JavaDocs to all relevant methods.
- [X] I have updated the documentation: 
  - User Guide Table of Contents, Description and Summary.
  - Developer Guide, if applicable.
- [X] I have ran `./gradlew test` or `./gradlew check` and all checks pass.
- [X] I have updated my project portfolio with what I have contributed.

### CODE REVIEW

File: src/main/java/seedu/address/model/person/Attendance.java [hunk 1]
```diff
@@ -1,59 +1,67 @@
...
         String[] ppArr = pp.split(",");
-        this.totalTut = atdArr.length;
         for (int i = 0; i < totalTut; i++) {
-            if (atdArr[i].equals("1")) {
-                this.attendanceList[i] = true;
-                this.participationList[i] = Integer.parseInt(ppArr[i].trim());
-            } else {
-                this.participationList[i] = 0;
+            attendanceList[i] = new AttendanceStatus(i);
+            participationList[i] = Integer.parseInt(ppArr[i]);
```
Comment (Choonyan02) [inline src/main/java/seedu/address/model/person/Attendance.java:31]: Why did you remove the defence against updating participation points when attendance isn't present/VR? Add it back in, we shouldn't always update participationList

File: src/main/java/seedu/address/model/person/Attendance.java [hunk 2]
```diff
@@ -1,59 +1,67 @@
...
     }
 
     /**
-     * Checks if a given week is a valid tutorial number (between 1 and 12).
-     *
+     * Checks if a given week is a valid tutorial number (between 1 and the maximum tutorial count).
      * @param tutorial A string representing a tutorial number.
      * @return `true` if the week is valid; otherwise, `false`.
      */
     public static boolean isValidTutorial(String tutorial) {
```
Comment (Choonyan02) [inline src/main/java/seedu/address/model/person/Attendance.java:52]: Not sure why you changed the method to a try catch when we are not throwing any errors. Does this count as abusing try catch? The previous method works fine.

**Review by Choonyan02 — CHANGES_REQUESTED** [review]
Mostly the same comments as the previous PR. Would be good to edit how AttendanceStatus works too, as mentioned in the previous PR, but that can be done in the future.

File: src/main/java/seedu/address/logic/commands/ClearCommand.java [hunk 3]
```diff
@@ -14,6 +17,8 @@ public class ClearCommand extends Command {
 
     @Override
     public CommandResult execute(Model model) {
+        requireNonNull(model);
+        model.setAddressBook(new NpcTrack());
```
Comment (Choonyan02) [inline src/main/java/seedu/address/logic/commands/ClearCommand.java:21]: I think that this shouldn't be here, this class only exist to send a warning message. If you add it back in, the user will clear everything without any secondary protection

**Review by Choonyan02 — CHANGES_REQUESTED** [review]

**Review by et-irl — COMMENTED** [review]
For future reference, please rename your pull request to the specific problem you are fixing with the pull request.

In the pull request description, please provide the exact context and list of things you have changed to make it easier for the reviewer.

You may notice that the number of things you changed is massive, such as updating the UI, updating all references to the `AddressBook` to `NpcTrack`, as well as attendance related commands and a lot more.
Then, the number of things you have to specify in the title and description will be unwieldy. This is intentional, and is made this way so you make incremental and small PRs which are easy to review and are containable.

As it is, it's extremely difficult to give a review of your code. Next time, be sure to make a branch for every single small feature you would like to change, and make a pull request associated with each feature you would like to change.

**Review by Vanessamae23 — CHANGES_REQUESTED** [review]
Not too sure what is the intended command with `markAtd 2 t/1` but there is no proper error message on the frontend to tell user what is wrong.

There is exception thrown though 

***Action*** 
image

*** Result ***
image

File: src/main/java/seedu/address/model/person/Attendance.java [hunk 4]
```diff
@@ -252,4 +302,85 @@ public String listParticipation() {
...
+         * Unmarks the instance of AttendanceStatus
+         */
+        public void unMark() {
+            this.isUnmarked = true;
+            this.isAbsent = false;
+            this.isVr = false;
+            this.isPresent = false;
+        }
+
+        public boolean isMarked() {
```
Comment (Vanessamae23) [inline src/main/java/seedu/address/model/person/Attendance.java:368]: Do we need Javadocs for this?

**Review by Vanessamae23 — COMMENTED** [review]
image

Is this behaviour 7/5 expected?

**Review by Choonyan02 — APPROVED** [review]

**Review by Vanessamae23 — APPROVED** [review]

File: docs/UserGuide.md [hunk 5]
```diff
@@ -654,10 +663,23 @@ followed by a quick summary guide of the Graphical User Interface (GUI)
 |---------------|----------------------------------------------------------------|
 | **Parameter** | Details about the student that will be included in the command |
 | **Command**   | Instructions that `npc_track` will execute                     |
+| **Attendance  | The student's attendance status. Attendance statuses can either|
+|   status**    | be "present", "absent", or "valid reason".                     |
```
Comment (et-irl) [inline docs/UserGuide.md:667]: This should be 1 block, not split across 2 rows.

File: src/main/java/seedu/address/model/person/Attendance.java [hunk 6]
```diff
@@ -67,23 +89,32 @@ public static boolean isValidParticipation(String pp) {
     }
 
     /**
-     * Returns the number of weeks present based on the attendance record.
-     *
-     * @return The count of weeks marked as present.
+     * Checks if the status string is a valid status.
+     * @param status The status string.
+     * @return 'true' if the status is valid, else 'false'.
```
Comment (et-irl) [inline src/main/java/seedu/address/model/person/Attendance.java:94]: Backticks should be used here.

File: docs/UserGuide.md [hunk 7]
```diff
@@ -158,9 +157,7 @@ Step 5 : Start using the app
 <a name="help"></a>
 ### Viewing help : `help`
 
-If you have trouble using `npc_track`, simply type the `help` command.
-
-Opens the user guide in your browser.
+If you have trouble using `npc_track`, simply type the `help` command. A link to this very user guide will be shown.
```
Comment (et-irl) [inline docs/UserGuide.md:160]: This is not true. The link to the user guide will not be shown, and it will be opened immediately.

File: docs/UserGuide.md [hunk 8]
```diff
@@ -26,9 +27,7 @@ We help teaching assistants to be able to bring together their teaching needs in
 adding extra information for a student named Ted can easily be done using the various keyword features as listed in 
 our [Features Section](#features)
 
-`npc_track` saves teaching assistant from having to spend so much precious time by reducing the complexity and the 
-need to 
-navigate different platforms just to search their students. Now, it can be done in a one-stop manner using `npc_track`
+`npc_track` saves teaching assistant from having to spend so much precious time by reducing the complexity by doing it in a one-stop manner using `npc_track`.
 
 Not ready to use `npc_track`? Fret Not, you can jump to the [Quick Start](#quick-start) section to begin your 
 `npc-track`
```
Comment (et-irl) [inline docs/UserGuide.md:33]: Do you think that it should be `npc_track`, not `npc-track`?

File: src/main/java/seedu/address/logic/commands/InputGroupParticipationCommand.java [hunk 9]
```diff
@@ -22,6 +22,9 @@ public class InputGroupParticipationCommand extends Command {
     public static final String SUCCESS_MSG = "Participation points input successfully!";
     public static final String ATTENDANCE_NOT_MARKED = "Before inputting participation points, "
             + "mark the attendance of the student first!";
+    public static final String ATTENDANCE_ABSENT = "A student was absent for this tutorial!";
+    public static final String PARTICIPATION_POINTS_OUT_OF_RANGE = "Participation points "
+            + "must be between 0 and 1000.";
```
Comment (et-irl) [inline src/main/java/seedu/address/logic/commands/InputGroupParticipationCommand.java:27]: Do you think that, for code maintainability, we should define it this way instead?

```java
    public static final String PARTICIPATION_POINTS_OUT_OF_RANGE = "Participation points "
            + "must be between 0 and " + MAXIMUM_PARTICIPATION_POINTS + ".";
```

File: src/main/java/seedu/address/logic/commands/InputParticipationCommand.java [hunk 10]
```diff
@@ -21,6 +21,9 @@ public class InputParticipationCommand extends Command {
     public static final String SUCCESS_MSG = "Participation points input successfully!";
     public static final String ATTENDANCE_NOT_MARKED = "Before inputting participation points, "
             + "mark the attendance of the student first!";
+    public static final String ATTENDANCE_ABSENT = "The student was absent for this tutorial!";
+    public static final String PARTICIPATION_POINTS_OUT_OF_RANGE = "Participation points "
+            + "must be between 0 and 1000.";
```
Comment (et-irl) [inline src/main/java/seedu/address/logic/commands/InputParticipationCommand.java:26]: Do you think that, for code maintainability, we should define it this way instead?

```java
    public static final String PARTICIPATION_POINTS_OUT_OF_RANGE = "Participation points "
            + "must be between 0 and " + MAXIMUM_PARTICIPATION_POINTS + ".";
```

File: src/main/java/seedu/address/logic/commands/MarkAttendanceCommand.java [hunk 11]
```diff
@@ -23,22 +24,24 @@ public class MarkAttendanceCommand extends Command {
...
     private final Index index;
     private final Index tut;
-
+    private final String status;
     /**
      * Constructs a new MarkAttendanceCommand to mark attendance for a student on a specific week.
      *
      * @param index The index of the student to mark attendance for.
      * @param tut The index of the week to mark attendance on.
      */
```
Comment (et-irl) [inline src/main/java/seedu/address/logic/commands/MarkAttendanceCommand.java:38]: Your JavaDocs are unupdated, to include `String status`.

File: src/main/java/seedu/address/logic/commands/MarkGroupAttendanceCommand.java [hunk 12]
```diff
@@ -23,22 +24,26 @@ public class MarkGroupAttendanceCommand extends Command {
...
     private final Group group;
 
+    private final String status;
+
     /**
      * Constructs a new MarkAttendanceCommand to mark attendance for a student on a specific week.
      *
      * @param group The group to mark group attendance for.
      * @param tut The index of the week to mark group attendance on.
      */
```
Comment (et-irl) [inline src/main/java/seedu/address/logic/commands/MarkGroupAttendanceCommand.java:40]: Your JavaDocs are unupdated, to include `String status`.

File: src/main/java/seedu/address/logic/parser/MarkAttendanceParser.java
Comment (et-irl) [inline src/main/java/seedu/address/logic/parser/MarkAttendanceParser.java:1]: We should be using the construct of checking if all prefixes are present much like in other commands for `MarkAttendanceParser.java` and `MarkGroupAttendanceParser.java`, instead of maintaining stateful variables with `status= = null`, as this can get unmaintainable and unreadable.

Comment (et-irl) [reply src/main/java/seedu/address/logic/parser/MarkAttendanceParser.java:1]: We can let this slide for now, as it works.

File: src/main/java/seedu/address/logic/parser/MarkGroupAttendanceParser.java [hunk 13]
```diff
@@ -37,9 +38,20 @@ public MarkGroupAttendanceCommand parse(String args) throws ParseException {
             tutorial = ParserUtil.parseTutorial(argMultimap.getValue(PREFIX_TUTORIAL).get());
         }
 
+        //parse status
```
Comment (et-irl) [inline src/main/java/seedu/address/logic/parser/MarkGroupAttendanceParser.java:41]: Nitpick, but there should be a space after `//`.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 14]
```diff
@@ -53,9 +53,24 @@ public static int parseTutorial(String tutorial) throws ParseException {
         return Attendance.convertToIntegerWeek(trimmedTutorial);
     }
 
+    /**
+     * Parses a {@code pp} into an {@code String} and returns it. Leading and trailing whitespaces will be trimmed.
+     * @param status String value of  status
```
Comment (et-irl) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:58]: Nitpick, but there shouldn't be 2 whitespaces in a row here.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 15]
```diff
@@ -53,9 +53,24 @@ public static int parseTutorial(String tutorial) throws ParseException {
...
+     * @return int value of status if valid
```
Comment (et-irl) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:59]: This is inaccurate, `int` is not returned.

File: src/main/java/seedu/address/model/person/Attendance.java [hunk 16]
```diff
@@ -1,58 +1,80 @@
...
+
+    private enum AttendanceStatus {
+        UNMARKED,
+        PRESENT,
+        ABSENT,
+        VR;
+
+        @Override
+        public String toString() {
+            return this == PRESENT ? "P" : this == ABSENT ? "A" : this == VR ? "VR" : "U";
```
Comment (et-irl) [inline src/main/java/seedu/address/model/person/Attendance.java:26]: This is unsound style, which is unreadable and unmaintainable. Would it be better to use `if-else` statements or even better, a `case` `switch` statement?

File: src/main/java/seedu/address/model/person/Attendance.java [hunk 17]
```diff
@@ -92,62 +123,92 @@ public int getTotalWeeks() {
...
 
     /**
      * Checks if a specific tutorial is marked as attended.
-     *
      * @param tutorial The week to check (1-12).
      * @return `true` if the week is marked as attended; otherwise, `false`.
      */
     public boolean isMarkedWeek(int tutorial) {
-        return this.attendanceList[tutorial];
+        return !(this.attendanceList[tutorial] == AttendanceStatus.UNMARKED);
```
Comment (et-irl) [inline src/main/java/seedu/address/model/person/Attendance.java:171]: Wouldn't it be cleaner to use `!==`?

File: src/main/java/seedu/address/model/util/SampleDataUtil.java
Comment (et-irl) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java:1]: This file is problematic. We have updated the tag sets and names to be saner:
- Tags represent actual classes
- Names are in alphabetical order
This way, the data is more organised and we can tell exactly when something goes wrong.

Don't you think we should keep that?

File: src/main/java/seedu/address/ui/PersonCard.java [hunk 18]
```diff
@@ -61,8 +63,10 @@ public PersonCard(Person person, int displayedIndex) {
         String telegramAddress = person.getTelegramHandle() != null ? " (@" + person.getTelegramHandle().value + ")"
                 : " -";
         name.setText(person.getName().fullName + telegramAddress);
-        attendance.setText(String.format("Attendance: %s",
-                this.person.getStyledAttendanceList("✓", "✕")));
+        attendance.setText(String.format("Attendance: %d/%d  Attendance summary: %s",
+                this.person.getAttendance().getWeeksPresent(), this.person.getAttendance().getTotalMarkedTut(),
+                this.person.getAttendance().getStyledStatusList("❌", "\uD83D\uDE4B\u200D♂\uFE0F",
+                        "\uD83D\uDE47\u200D♂\uFE0F", "?")));
```
Comment (et-irl) [inline src/main/java/seedu/address/ui/PersonCard.java:69]: What are these symbols? Would it be better to leave a comment here explaining what `"\uD83D\uDE4B\u200D♂\uFE0F"` and `"\uD83D\uDE47\u200D♂\uFE0F"` are?

File: src/main/java/seedu/address/ui/PersonCard.java [hunk 19]
```diff
@@ -61,8 +63,10 @@ public PersonCard(Person person, int displayedIndex) {
         String telegramAddress = person.getTelegramHandle() != null ? " (@" + person.getTelegramHandle().value + ")"
                 : " -";
         name.setText(person.getName().fullName + telegramAddress);
-        attendance.setText(String.format("Attendance: %s",
-                this.person.getStyledAttendanceList("✓", "✕")));
+        attendance.setText(String.format("Attendance: %d/%d  Attendance summary: %s",
```
Comment (et-irl) [inline src/main/java/seedu/address/ui/PersonCard.java:66]: Is this a bit terse? It is obvious from context that the following `%s` is a summary - shouldn't we omit `Attendance summary` in favour of a more minimalistic display?

File: src/main/resources/view/PersonListCard.fxml [hunk 20]
```diff
@@ -29,24 +29,24 @@
...
+        <Region HBox.hgrow="ALWAYS" />
       </HBox>
       <VBox spacing="3">
         <FlowPane hgap="5" alignment="CENTER_LEFT">
           <Label styleClass="cell_small_label_tags" text="🏷 Tags:" prefWidth="175" />
           <FlowPane fx:id="tags" />
         </FlowPane>
         <FlowPane hgap="5" alignment="CENTER_LEFT">
-          <Label styleClass="cell_small_label_tags" text="🗒 Comments:" prefWidth="175" />
+          <Label styleClass="cell_small_label_tags" text="🗒️ Comments:" prefWidth="200" />
```
Comment (et-irl) [inline src/main/resources/view/PersonListCard.fxml:41]: We should not re-introduce the regression of emotes causing display issues on Windows. While the `prefWidth` can change, we should not change the text representation.

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 21]
```diff
@@ -31,62 +28,61 @@ public class TypicalPersons {
...
-            .withAttendance(VALID_ATTENDANCE_AMY, VALID_PART_AMY).build();
+            .withAttendance(Attendance.ORIGINAL_ATD, Attendance.ORIGINAL_PART).build();
     public static final Person BOB = new PersonBuilder().withName(VALID_NAME_BOB).withPhone(VALID_PHONE_BOB)
             .withEmail(VALID_EMAIL_BOB).withTelegram(VALID_TELEGRAM_BOB).withTags(VALID_TAG_HUSBAND, VALID_TAG_FRIEND)
-            .withAttendance(VALID_ATTENDANCE_BOB, VALID_PART_BOB).build();
+            .withAttendance(Attendance.ORIGINAL_ATD, Attendance.ORIGINAL_PART).build();
 
 
-    public static final String KEYWORD_MATCHING_MEIER = "Meier";
+    public static final String KEYWORD_MATCHING_MEIER = "Meier"; // A keyword that matches MEIER
```
Comment (et-irl) [inline src/test/java/seedu/address/testutil/TypicalPersons.java:76]: This comment is not useful. We should not include this comment.

File: src/test/java/seedu/address/testutil/TypicalPersons.java [hunk 22]
```diff
@@ -31,62 +28,61 @@ public class TypicalPersons {
...
 
     private TypicalPersons() {
-        // prevents instantiation
-    }
+    } // prevents instantiation
```
Comment (et-irl) [inline src/test/java/seedu/address/testutil/TypicalPersons.java:79]: Is there a reason for preferring this comment style in the constructor? A comment after a closing bracket `}` feels unintuitive.

File: src/test/java/seedu/address/testutil/TypicalPersons.java
Comment (et-irl) [inline src/test/java/seedu/address/testutil/TypicalPersons.java:1]: The person names should not be overridden in this file. We have changed the data to be sane, such as making it in alphabetical order, much like the previous comment I have added in the review. Please do not change this data.

**Review by et-irl — CHANGES_REQUESTED** [review]
Many issues to be rectified before we can merge this - do not put so many features in one pull request in the future.

File: docs/UserGuide.md (hunk 8 above)
Comment (et-irl) [reply docs/UserGuide.md:33]: Nitpick: saves teaching **assistants**, not assistant.

File: src/main/java/seedu/address/ui/PersonCard.java (hunk 18 above)
Comment (proto-aiken-13) [reply src/main/java/seedu/address/ui/PersonCard.java:69]: Resolved in the updated PR by declaring them as String variables within the PersonCard class.

**Review by et-irl — CHANGES_REQUESTED** [review]
A few minor nits based on scrutinising the codebase itself.

Could you elaborate, (in the PR description next time), exactly what features you are adding and how you expect them to work?

When we mark attendances, the capacity seems to grow in accordance. To confirm, is this the intended behaviour? 

The emoji for the attendance list does not seem to be rendering correctly. Is this a Linux issue? If it works on Windows, it is okay.

image

File: src/main/java/seedu/address/logic/parser/MarkAttendanceParser.java [hunk 23]
```diff
@@ -21,10 +22,11 @@ public class MarkAttendanceParser implements Parser<MarkAttendanceCommand> {
      */
     public MarkAttendanceCommand parse(String args) throws ParseException {
         requireNonNull(args);
-        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_TUTORIAL);
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args, PREFIX_TUTORIAL, PREFIX_PARTICIPATION_STATUS);
 
         Index index;
         int tutorial = 0;
+        String status = null;
```
Comment (Kurtyjlee) [inline src/main/java/seedu/address/logic/parser/MarkAttendanceParser.java:29]: Might be safer to create a status class so that you can use the properties of an empty status :))

File: src/main/java/seedu/address/logic/parser/MarkAttendanceParser.java [hunk 24]
```diff
@@ -36,10 +38,18 @@ public MarkAttendanceCommand parse(String args) throws ParseException {
...
+        if (argMultimap.getValue(PREFIX_PARTICIPATION_STATUS).isPresent()) {
+            status = ParserUtil.parseParticipationStatus(argMultimap.getValue(PREFIX_PARTICIPATION_STATUS).get());
+        }
+
         if (!(tutorial >= 1 && tutorial <= 12)) {
             throw new ParseException(Attendance.TUTORIAL_ERROR_MSG);
         }
 
-        return new MarkAttendanceCommand(index, Index.fromOneBased(tutorial));
+        if (status == null) {
```
Comment (Kurtyjlee) [inline src/main/java/seedu/address/logic/parser/MarkAttendanceParser.java:49]: This null check would be safer with an empty Status class :))

File: src/main/java/seedu/address/model/person/Attendance.java [hunk 25]
```diff
@@ -1,58 +1,97 @@
...
+    public static final String STATUS_ERROR_MSG = "Status is invalid! It should either be P, VR or A.";
+    public static final String ORIGINAL_ATD = "U,U,U,U,U,U,U,U,U,U,U,U";
     public static final String PARTICIPATION_ERROR_MSG = "Please input a small non-negative number.";
-    public static final String ORIGINAL_ATD = "0,0,0,0,0,0,0,0,0,0,0,0";
     public static final String ORIGINAL_PART = "0,0,0,0,0,0,0,0,0,0,0,0";
-    private int totalTut;
-    private final boolean[] attendanceList = new boolean[12];
-    private final int[] participationList = new int[12];
+    private static int totalTut = 12;
+    private final int[] participationList = new int[totalTut];
```
Comment (Kurtyjlee) [inline src/main/java/seedu/address/model/person/Attendance.java:15]: This can be a class as well so that you can store all methods for this list and make everything neater. 😀

File: src/main/java/seedu/address/model/person/Attendance.java [hunk 26]
```diff
@@ -1,58 +1,97 @@
...
         String[] ppArr = pp.split(",");
-        this.totalTut = atdArr.length;
         for (int i = 0; i < totalTut; i++) {
-            if (atdArr[i].equals("1")) {
-                this.attendanceList[i] = true;
-                this.participationList[i] = Integer.parseInt(ppArr[i].trim());
-            } else {
-                this.participationList[i] = 0;
+            switch (atdArr[i]) {
+            case "A":
```
Comment (Kurtyjlee) [inline src/main/java/seedu/address/model/person/Attendance.java:60]: Should this an enum? Asking for a friend 🤗

File: src/main/java/seedu/address/model/person/Attendance.java [hunk 27]
```diff
@@ -1,58 +1,97 @@
...
 
     /**
-     * Checks if a given week is a valid tutorial number (between 1 and 12).
-     *
+     * Checks if a given week is a valid tutorial number (between 1 and the maximum tutorial count).
      * @param tutorial A string representing a tutorial number.
      * @return `true` if the week is valid; otherwise, `false`.
      */
     public static boolean isValidTutorial(String tutorial) {
         if (!tutorial.matches("[0-9]+")) {
```
Comment (Kurtyjlee) [inline src/main/java/seedu/address/model/person/Attendance.java:85]: This regex looks like it can be a constant!

**Review by Kurtyjlee — COMMENTED** [review]
Just a fellow CS2103T passerby with a few comments 😋

**Review by et-irl — APPROVED** [review]
Looks good to me.
